File: mathematical-art/flat-geometry/julia/flat_geometry.jl
# FLAT-GEOMETRY-V1
# Numerical geometry mirror: Julia
using LinearAlgebra
using Printf

regular_polygon(n::Integer, c::Tuple{<:Real,<:Real}, r::Real) = [
    (c[1] + r*cos(2π*j/n + π/2), c[2] + r*sin(2π*j/n + π/2))
    for j in 0:n-1
]

function hyperplane_distance(x::AbstractVector, a::AbstractVector, b::Real)
    @assert length(x) == length(a)
    na = norm(a)
    @assert na > 0 "a must be nonzero"
    return abs(dot(a, x) - b) / na
end

struct AffineFlat
    x0::Vector{Float64}
    basis::Matrix{Float64}
    function AffineFlat(x0, basis)
        B = Matrix{Float64}(basis)
        rank(B) == size(B, 2) || error("basis vectors must be linearly independent")
        length(x0) == size(B, 1) || error("ambient dimensions must agree")
        new(Vector{Float64}(x0), B)
    end
end

dimension(F::AffineFlat) = size(F.basis, 2)
point(F::AffineFlat, λ::AbstractVector) = F.x0 + F.basis*λ

for n in 3:10
    pts = regular_polygon(n, (0.0, 0.0), 1.0)
    @printf("%2d-gon: first vertex = (%.6f, %.6f)\n", n, pts[1]...)
end

File: mathematical-art/flat-geometry/csharp/FlatGeometry.cs
// FLAT-GEOMETRY-V1
// Engineering integration mirror: C# / SVG.
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

public readonly record struct Point2(double X, double Y);

public static class FlatGeometry
{
    public static IReadOnlyList<Point2> RegularPolygon(int n, double cx, double cy, double r)
    {
        if (n < 3) throw new ArgumentOutOfRangeException(nameof(n));
        return Enumerable.Range(0, n).Select(j =>
        {
            var t = Math.PI / 2.0 + 2.0 * Math.PI * j / n;
            return new Point2(cx + r * Math.Cos(t), cy - r * Math.Sin(t));
        }).ToArray();
    }

    public static double HyperplaneDistance(double[] x, double[] a, double b)
    {
        if (x.Length != a.Length) throw new ArgumentException("ambient dimensions must agree");
        var dot = x.Zip(a, (xi, ai) => xi * ai).Sum();
        var norm = Math.Sqrt(a.Sum(ai => ai * ai));
        if (norm == 0) throw new ArgumentException("a must be nonzero");
        return Math.Abs(dot - b) / norm;
    }

    public static string PolygonSvg(int n, double cx, double cy, double r)
    {
        var points = string.Join(" ", RegularPolygon(n, cx, cy, r)
            .Select(p => string.Create(CultureInfo.InvariantCulture, $"{p.X:F2},{p.Y:F2}")));
        return $"<polygon points=\"{points}\" fill=\"none\" stroke=\"#6D28D9\"/>";
    }
}
